qiskit/algorithms/minimum_eigen_solvers/vqe.py

```diff
@@ -454,19 +464,24 @@ def compute_minimum_eigenvalue(
         initial_point = _validate_initial_point(self.initial_point, self.ansatz)
 
         bounds = _validate_bounds(self.ansatz)
-
-        # We need to handle the array entries being Optional i.e. having value None
+        # We need to handle the array entries being zero or Optional i.e. having value None
         if aux_operators:
             zero_op = I.tensorpower(operator.num_qubits) * 0.0
-            converted = []
-            for op in aux_operators:
-                if op is None:
-                    converted.append(zero_op)
-                else:
-                    converted.append(op)
 
-            # For some reason Chemistry passes aux_ops with 0 qubits and paulis sometimes.
-            aux_operators = [zero_op if op == 0 else op for op in converted]
+            # Convert the None and zero values when aux_operators is a list.
+            # Drop None and zero values when aux_operators is a dict.
```

Comment:
I am not sure I fully follow this comment. In the past we either had an operator instance or None. None was the case where we had to drop an operator from the list during symmetry processing as it did not commute and therefore any measurement with the final state of the main operator was meaningless. The None was put there to maintain the list order/indexes with None returned as a result such that the caller knew this has occurred. Now a valid instance could also end up being empty so these were handled explicitly to return 0.

In looking at the dict the intent was for the user/higher-level code, to pass in a dictionary such that non-commuting operators would just be dropped and the return result would lack just the key; So if I pass in a aux_ops as a set of key value pairs I go through the results according to my keys and if a key is missing I know it could not be measured. An empty operator should still return a 0 value though.

Reply:
I see, I wrongly assumed that None and 0 were equivalent since they both got converted to zero operators in the list. I'll make it so that the dict only drops None values and keeps the zero operators

Reply:
In terms of None the idea is that only valid (potentially they might be empty instances though) should be passed through with the dictionary. Maybe if it is None then a warning should be logged noting that its dropping an invalid dictionary entry since its value is None (unlike the list we do not intend to give back a key with a None value so passing in None as a value is not really right in my mind)

(How things currently work in the higher level stack is given we do symmetry reduction on the main operator the aux_ops we have we need to check they commute - any that didn't in the current list mechanism were replaced by None to preserve order and pass None back to indicate this back up the stack to the originator of the aux ops that the order is preserved and they know which value goes with what - a check for None is needed there which means the operator could not be measured. What we are looking at going forwards is simply that the originator sets up a dictionary and then accesses the results via the same keys - they need to check if a key is missing, which means the operator was not able to be measured. So rather than the stack putting None into the aux_ops list to preserve order as the list is processed, the thinking is that the stack will simply drop the key/value pair from the dictionary - i.e. not replace its value by None and preserve the key)